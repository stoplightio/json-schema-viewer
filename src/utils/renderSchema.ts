import { JSONSchema4 } from 'json-schema';
import _cloneDeep = require('lodash/cloneDeep');
import _isEmpty = require('lodash/isEmpty');
import _merge = require('lodash/merge');
import { IArrayNode, IObjectNode, ITreeNodeMeta, SchemaKind, SchemaTreeListNode } from '../types';
import { DIVIDERS } from './dividers';
import { getPrimaryType } from './getPrimaryType';
import { isCombiner } from './isCombiner';
import { isRef } from './isRef';
import { walk } from './walk';

// @ts-ignore no typings
import * as resolveAllOf from 'json-schema-merge-allof';

type Walker = (
  schema: JSONSchema4,
  level?: number,
  meta?: ITreeNodeMeta,
  options?: { mergeAllOf?: boolean },
) => IterableIterator<SchemaTreeListNode>;

const getProperties: Walker = function*(schema, level = 0, meta) {
  if (schema.properties !== undefined) {
    const { path } = meta!;
    for (const [prop, property] of Object.entries(schema.properties)) {
      yield* renderSchema(property, level + 1, {
        name: prop,
        required: Array.isArray(schema.required) && schema.required.includes(prop),
        path: [...path, prop],
      });
    }
  }
};

const getPatternProperties: Walker = function*(schema, level = 0, meta) {
  if (schema.patternProperties !== undefined) {
    const { path } = meta!;
    for (const [prop, property] of Object.entries(schema.patternProperties)) {
      yield* renderSchema(property, level + 1, {
        name: prop,
        path: [...path, prop],
        pattern: true,
      });
    }
  }
};

export const renderSchema: Walker = function*(schema, level = 0, meta = { path: [] }, options = { mergeAllOf: false }) {
  if (typeof schema !== 'object' || schema === null) {
    throw new TypeError(
      `Expected schema to be an "object" but received ${schema === null ? '"null"' : `a "${typeof schema}"`}`,
    );
  }

  const resolvedSchema = _cloneDeep(schema);
  const parsedSchema = options.mergeAllOf
    ? resolveAllOf(resolvedSchema, {
        resolvers: {
          defaultResolver(values: any) {
            // Handle merging unknown properties
            // @ts-ignore not accepting values
            return _merge(...values);
          },
        },
      })
    : resolvedSchema;

  const { path } = meta;

  // TODO: make use of mergeAllOf (is this needed?)
  for (const node of walk(resolvedSchema)) {
    const baseNode: SchemaTreeListNode = {
      id: node.id,
      level,
      name: '',
      metadata: {
        ...node,
        ...meta,
        ...(parsedSchema.items !== undefined &&
          !Array.isArray(parsedSchema.items) && {
            subtype: '$ref' in parsedSchema.items ? `$ref( ${parsedSchema.items.$ref} )` : parsedSchema.items.type,
          }),
        path,
      },
    };

    if (isRef(node)) {
      // we expect the schema to be dereferenced
      // const resolved = lookupRef(path, dereferencedSchema);
      // if (resolved) {
      //   yield* renderSchema(
      //     resolved,
      //     level,
      //     {
      //       ...meta,
      //       inheritedFrom: node.$ref,
      //     },
      //     { mergeAllOf: false },
      //   );
      // }
      yield {
        ...baseNode,
        metadata: {
          ...baseNode.metadata,
          $ref: node.$ref,
        },
      } as SchemaTreeListNode;
    } else if (isCombiner(node)) {
      yield {
        ...baseNode,
        canHaveChildren: true,
      };

      if (node.properties !== undefined) {
        for (const [i, property] of node.properties.entries()) {
          if ('type' in node) {
            property.type = property.type || node.type;
          }

          yield* renderSchema(property, level + 1, {
            ...(i !== 0 && { divider: DIVIDERS[node.combiner] }),
            path: [...path, schema.allOf ? 'allOf' : 'properties', i],
          });
        }
      }
    } else {
      switch (getPrimaryType(node)) {
        case SchemaKind.Array:
          yield {
            ...baseNode,
            ...('items' in node &&
              !_isEmpty(node.items) &&
              (baseNode.metadata!.subtype === 'object' || Array.isArray(node.items)) && { canHaveChildren: true }),
            metadata: {
              ...baseNode.metadata,
              // https://tools.ietf.org/html/draft-fge-json-schema-validation-00#section-5.3.1.2
              ...(!('subtype' in baseNode) &&
                (node as IArrayNode).additionalItems && { additional: (node as IArrayNode).additionalItems }),
            },
          } as SchemaTreeListNode;

          if (Array.isArray(parsedSchema.items)) {
            for (const [i, property] of parsedSchema.items.entries()) {
              yield* renderSchema(property, level + 1, {
                path: [...path, 'items', i],
              });
            }
          } else if (parsedSchema.items) {
            switch (baseNode.metadata && baseNode.metadata.subtype) {
              case SchemaKind.Object:
                yield* getProperties(parsedSchema.items, level, {
                  ...meta,
                  path: [...path, 'items'],
                });
                break;
              case SchemaKind.Array:
                yield* renderSchema(parsedSchema.items, level, {
                  path,
                });
                break;
            }
          }

          break;
        case SchemaKind.Object:
          yield {
            ...baseNode,
            ...('properties' in node && !_isEmpty(node.properties) && { canHaveChildren: true }),
            metadata: {
              ...baseNode.metadata,
              ...((node as IObjectNode).additionalProperties && {
                additional: (node as IObjectNode).additionalProperties,
              }),
            },
          } as SchemaTreeListNode;

          yield* getProperties(parsedSchema, level, {
            path: [...path, 'properties'],
          });

          yield* getPatternProperties(parsedSchema, level, {
            path: [...path, 'patternProperties'],
          });

          break;
        default:
          yield baseNode;
      }
    }
  }
};
