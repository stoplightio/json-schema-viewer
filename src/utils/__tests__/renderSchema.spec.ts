import * as fs from 'fs';
import * as path from 'path';
import { renderSchema } from '../renderSchema';

const BASE_PATH = path.resolve(__dirname, '../../__fixtures__/');

jest.mock('../generateId', () => ({
  generateId: jest.fn(() => 'random-id'),
}));

describe('renderSchema util', () => {
  it.each([
    ['default-schema.json', ''],
    ['ref/original.json', 'ref/resolved.json'],
    ['combiner-schema.json', ''],
    ['array-of-objects.json', ''],
    ['array-of-refs.json', ''],
    ['array-of-allofs.json', ''],
    ['tickets.schema.json', ''],
  ])('should match %s', (schema, dereferenced) => {
    expect(
      Array.from(renderSchema(JSON.parse(fs.readFileSync(path.resolve(BASE_PATH, schema), 'utf8')))),
    ).toMatchSnapshot();
  });

  it('given schema with complex types, throws', () => {
    expect(() =>
      Array.from(
        renderSchema({
          type: [
            'null',
            {
              type: 'object',
              properties: {
                taskId: {
                  type: 'string',
                  format: 'uuid',
                },
              },
            },
          ],
        } as any),
      ),
    ).toThrow(
      'The "type" property must be a string, or an array of strings. Objects and array of objects are not valid.',
    );
  });
});
