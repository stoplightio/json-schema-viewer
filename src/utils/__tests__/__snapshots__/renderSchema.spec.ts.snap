// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderSchema util should match array-of-objects.json 1`] = `
Array [
  Object {
    "canHaveChildren": true,
    "id": "random-id",
    "level": 0,
    "metadata": Object {
      "additionalProperties": undefined,
      "annotations": Object {},
      "enum": undefined,
      "id": "random-id",
      "path": Array [],
      "patternProperties": undefined,
      "properties": Object {
        "propertyIsArrayOfObjects": Object {
          "items": Object {
            "properties": Object {
              "ArrayObjectProperty": Object {
                "type": "string",
              },
            },
            "type": "object",
          },
          "type": Array [
            "array",
          ],
        },
      },
      "type": "object",
      "validations": Object {},
    },
    "name": "",
  },
  Object {
    "id": "random-id",
    "level": 1,
    "metadata": Object {
      "additionalItems": undefined,
      "annotations": Object {},
      "enum": undefined,
      "id": "random-id",
      "items": undefined,
      "name": "propertyIsArrayOfObjects",
      "path": Array [
        "properties",
        "propertyIsArrayOfObjects",
      ],
      "required": false,
      "subtype": "object",
      "type": Array [
        "array",
      ],
      "validations": Object {},
    },
    "name": "",
  },
  Object {
    "id": "random-id",
    "level": 3,
    "metadata": Object {
      "annotations": Object {},
      "enum": undefined,
      "id": "random-id",
      "name": "ArrayObjectProperty",
      "path": Array [
        "properties",
        "propertyIsArrayOfObjects",
        "items",
        "ArrayObjectProperty",
      ],
      "required": false,
      "type": "string",
      "validations": Object {},
    },
    "name": "",
  },
]
`;

exports[`renderSchema util should match array-of-refs.json 1`] = `
Array [
  Object {
    "id": "random-id",
    "level": 0,
    "metadata": Object {
      "additionalItems": undefined,
      "annotations": Object {},
      "enum": undefined,
      "id": "random-id",
      "items": undefined,
      "path": Array [],
      "subtype": "$ref( ./models/todo-full.json )",
      "type": "array",
      "validations": Object {},
    },
    "name": "",
  },
]
`;

exports[`renderSchema util should match combiner-schema.json 1`] = `
Array [
  Object {
    "canHaveChildren": true,
    "id": "random-id",
    "level": 0,
    "metadata": Object {
      "annotations": Object {
        "title": "Todo Full",
      },
      "combiner": "allOf",
      "id": "random-id",
      "path": Array [],
      "properties": Array [
        Object {
          "properties": Object {
            "test": Object {
              "type": "string",
            },
          },
          "type": "object",
        },
        Object {
          "properties": Object {
            "completed_at": Object {
              "format": "date-time",
              "type": Array [
                "string",
                "null",
              ],
            },
            "created_at": Object {
              "format": "date-time",
              "type": "string",
            },
            "id": Object {
              "maximum": 1000000,
              "minimum": 0,
              "type": "integer",
            },
            "updated_at": Object {
              "format": "date-time",
              "type": "string",
            },
          },
          "required": Array [
            "id",
          ],
          "type": "object",
        },
      ],
      "type": "object",
    },
    "name": "",
  },
  Object {
    "canHaveChildren": true,
    "id": "random-id",
    "level": 1,
    "metadata": Object {
      "additionalProperties": undefined,
      "annotations": Object {},
      "enum": undefined,
      "id": "random-id",
      "path": Array [
        "properties",
        0,
      ],
      "patternProperties": undefined,
      "properties": Object {
        "test": Object {
          "type": "string",
        },
      },
      "type": "object",
      "validations": Object {},
    },
    "name": "",
  },
  Object {
    "id": "random-id",
    "level": 2,
    "metadata": Object {
      "annotations": Object {},
      "enum": undefined,
      "id": "random-id",
      "name": "test",
      "path": Array [
        "properties",
        0,
        "properties",
        "test",
      ],
      "required": false,
      "type": "string",
      "validations": Object {},
    },
    "name": "",
  },
  Object {
    "canHaveChildren": true,
    "id": "random-id",
    "level": 1,
    "metadata": Object {
      "additionalProperties": undefined,
      "annotations": Object {},
      "divider": "and",
      "enum": undefined,
      "id": "random-id",
      "path": Array [
        "properties",
        1,
      ],
      "patternProperties": undefined,
      "properties": Object {
        "completed_at": Object {
          "format": "date-time",
          "type": Array [
            "string",
            "null",
          ],
        },
        "created_at": Object {
          "format": "date-time",
          "type": "string",
        },
        "id": Object {
          "maximum": 1000000,
          "minimum": 0,
          "type": "integer",
        },
        "updated_at": Object {
          "format": "date-time",
          "type": "string",
        },
      },
      "type": "object",
      "validations": Object {},
    },
    "name": "",
  },
  Object {
    "id": "random-id",
    "level": 2,
    "metadata": Object {
      "annotations": Object {},
      "enum": undefined,
      "id": "random-id",
      "name": "id",
      "path": Array [
        "properties",
        1,
        "properties",
        "id",
      ],
      "required": true,
      "type": "integer",
      "validations": Object {},
    },
    "name": "",
  },
  Object {
    "id": "random-id",
    "level": 2,
    "metadata": Object {
      "annotations": Object {},
      "enum": undefined,
      "id": "random-id",
      "name": "completed_at",
      "path": Array [
        "properties",
        1,
        "properties",
        "completed_at",
      ],
      "required": false,
      "type": Array [
        "string",
        "null",
      ],
      "validations": Object {
        "format": "date-time",
      },
    },
    "name": "",
  },
  Object {
    "id": "random-id",
    "level": 2,
    "metadata": Object {
      "annotations": Object {},
      "enum": undefined,
      "id": "random-id",
      "name": "created_at",
      "path": Array [
        "properties",
        1,
        "properties",
        "created_at",
      ],
      "required": false,
      "type": "string",
      "validations": Object {
        "format": "date-time",
      },
    },
    "name": "",
  },
  Object {
    "id": "random-id",
    "level": 2,
    "metadata": Object {
      "annotations": Object {},
      "enum": undefined,
      "id": "random-id",
      "name": "updated_at",
      "path": Array [
        "properties",
        1,
        "properties",
        "updated_at",
      ],
      "required": false,
      "type": "string",
      "validations": Object {
        "format": "date-time",
      },
    },
    "name": "",
  },
]
`;

exports[`renderSchema util should match default-schema.json 1`] = `
Array [
  Object {
    "canHaveChildren": true,
    "id": "random-id",
    "level": 0,
    "metadata": Object {
      "additionalProperties": undefined,
      "annotations": Object {
        "title": "User",
      },
      "enum": undefined,
      "id": "random-id",
      "path": Array [],
      "patternProperties": Object {
        "^id_": Object {
          "type": "number",
        },
        "_name$": Object {
          "type": "string",
        },
        "foo": Object {
          "type": "integer",
        },
      },
      "properties": Object {
        "age": Object {
          "maximum": 150,
          "minimum": 0,
          "type": "number",
        },
        "completed_at": Object {
          "format": "date-time",
          "type": "string",
        },
        "email": Object {
          "format": "email",
          "minLength": 2,
          "type": "string",
        },
        "items": Object {
          "items": Object {
            "type": Array [
              "string",
              "number",
            ],
          },
          "type": Array [
            "null",
            "array",
          ],
        },
        "name": Object {
          "description": "The user's full name.",
          "type": "string",
        },
        "permissions": Object {
          "properties": Object {
            "ids": Object {
              "items": Object {
                "type": "integer",
              },
              "type": "array",
            },
          },
          "type": Array [
            "string",
            "object",
          ],
        },
        "plan": Object {
          "anyOf": Array [
            Object {
              "properties": Object {
                "bar": Object {
                  "type": "string",
                },
                "foo": Object {
                  "type": "string",
                },
              },
              "required": Array [
                "foo",
                "bar",
              ],
              "type": "object",
            },
            Object {
              "items": Object {
                "type": "integer",
              },
              "type": "array",
            },
          ],
        },
        "ref": Object {
          "$ref": "#/definitions/error-response",
        },
      },
      "type": "object",
      "validations": Object {},
    },
    "name": "",
  },
  Object {
    "id": "random-id",
    "level": 1,
    "metadata": Object {
      "annotations": Object {
        "description": "The user's full name.",
      },
      "enum": undefined,
      "id": "random-id",
      "name": "name",
      "path": Array [
        "properties",
        "name",
      ],
      "required": true,
      "type": "string",
      "validations": Object {},
    },
    "name": "",
  },
  Object {
    "id": "random-id",
    "level": 1,
    "metadata": Object {
      "annotations": Object {},
      "enum": undefined,
      "id": "random-id",
      "name": "age",
      "path": Array [
        "properties",
        "age",
      ],
      "required": true,
      "type": "number",
      "validations": Object {
        "maximum": 150,
        "minimum": 0,
      },
    },
    "name": "",
  },
  Object {
    "id": "random-id",
    "level": 1,
    "metadata": Object {
      "annotations": Object {},
      "enum": undefined,
      "id": "random-id",
      "name": "completed_at",
      "path": Array [
        "properties",
        "completed_at",
      ],
      "required": true,
      "type": "string",
      "validations": Object {
        "format": "date-time",
      },
    },
    "name": "",
  },
  Object {
    "id": "random-id",
    "level": 1,
    "metadata": Object {
      "additionalItems": undefined,
      "annotations": Object {},
      "enum": undefined,
      "id": "random-id",
      "items": undefined,
      "name": "items",
      "path": Array [
        "properties",
        "items",
      ],
      "required": false,
      "subtype": Array [
        "string",
        "number",
      ],
      "type": Array [
        "null",
        "array",
      ],
      "validations": Object {},
    },
    "name": "",
  },
  Object {
    "id": "random-id",
    "level": 1,
    "metadata": Object {
      "annotations": Object {},
      "enum": undefined,
      "id": "random-id",
      "name": "email",
      "path": Array [
        "properties",
        "email",
      ],
      "required": false,
      "type": "string",
      "validations": Object {
        "format": "email",
        "minLength": 2,
      },
    },
    "name": "",
  },
  Object {
    "canHaveChildren": true,
    "id": "random-id",
    "level": 1,
    "metadata": Object {
      "annotations": Object {},
      "combiner": "anyOf",
      "id": "random-id",
      "name": "plan",
      "path": Array [
        "properties",
        "plan",
      ],
      "properties": Array [
        Object {
          "properties": Object {
            "bar": Object {
              "type": "string",
            },
            "foo": Object {
              "type": "string",
            },
          },
          "required": Array [
            "foo",
            "bar",
          ],
          "type": "object",
        },
        Object {
          "items": Object {
            "type": "integer",
          },
          "type": "array",
        },
      ],
      "required": false,
    },
    "name": "",
  },
  Object {
    "canHaveChildren": true,
    "id": "random-id",
    "level": 2,
    "metadata": Object {
      "additionalProperties": undefined,
      "annotations": Object {},
      "enum": undefined,
      "id": "random-id",
      "path": Array [
        "properties",
        "plan",
        "properties",
        0,
      ],
      "patternProperties": undefined,
      "properties": Object {
        "bar": Object {
          "type": "string",
        },
        "foo": Object {
          "type": "string",
        },
      },
      "type": "object",
      "validations": Object {},
    },
    "name": "",
  },
  Object {
    "id": "random-id",
    "level": 3,
    "metadata": Object {
      "annotations": Object {},
      "enum": undefined,
      "id": "random-id",
      "name": "foo",
      "path": Array [
        "properties",
        "plan",
        "properties",
        0,
        "properties",
        "foo",
      ],
      "required": true,
      "type": "string",
      "validations": Object {},
    },
    "name": "",
  },
  Object {
    "id": "random-id",
    "level": 3,
    "metadata": Object {
      "annotations": Object {},
      "enum": undefined,
      "id": "random-id",
      "name": "bar",
      "path": Array [
        "properties",
        "plan",
        "properties",
        0,
        "properties",
        "bar",
      ],
      "required": true,
      "type": "string",
      "validations": Object {},
    },
    "name": "",
  },
  Object {
    "id": "random-id",
    "level": 2,
    "metadata": Object {
      "additionalItems": undefined,
      "annotations": Object {},
      "divider": "and/or",
      "enum": undefined,
      "id": "random-id",
      "items": undefined,
      "path": Array [
        "properties",
        "plan",
        "properties",
        1,
      ],
      "subtype": "integer",
      "type": "array",
      "validations": Object {},
    },
    "name": "",
  },
  Object {
    "canHaveChildren": true,
    "id": "random-id",
    "level": 1,
    "metadata": Object {
      "additionalProperties": undefined,
      "annotations": Object {},
      "enum": undefined,
      "id": "random-id",
      "name": "permissions",
      "path": Array [
        "properties",
        "permissions",
      ],
      "patternProperties": undefined,
      "properties": Object {
        "ids": Object {
          "items": Object {
            "type": "integer",
          },
          "type": "array",
        },
      },
      "required": false,
      "type": Array [
        "string",
        "object",
      ],
      "validations": Object {},
    },
    "name": "",
  },
  Object {
    "id": "random-id",
    "level": 2,
    "metadata": Object {
      "additionalItems": undefined,
      "annotations": Object {},
      "enum": undefined,
      "id": "random-id",
      "items": undefined,
      "name": "ids",
      "path": Array [
        "properties",
        "permissions",
        "properties",
        "ids",
      ],
      "required": false,
      "subtype": "integer",
      "type": "array",
      "validations": Object {},
    },
    "name": "",
  },
  Object {
    "id": "random-id",
    "level": 1,
    "metadata": Object {
      "$ref": "#/definitions/error-response",
      "id": "random-id",
      "name": "ref",
      "path": Array [
        "properties",
        "ref",
      ],
      "required": false,
    },
    "name": "",
  },
  Object {
    "id": "random-id",
    "level": 1,
    "metadata": Object {
      "annotations": Object {},
      "enum": undefined,
      "id": "random-id",
      "name": "^id_",
      "path": Array [
        "patternProperties",
        "^id_",
      ],
      "pattern": true,
      "type": "number",
      "validations": Object {},
    },
    "name": "",
  },
  Object {
    "id": "random-id",
    "level": 1,
    "metadata": Object {
      "annotations": Object {},
      "enum": undefined,
      "id": "random-id",
      "name": "foo",
      "path": Array [
        "patternProperties",
        "foo",
      ],
      "pattern": true,
      "type": "integer",
      "validations": Object {},
    },
    "name": "",
  },
  Object {
    "id": "random-id",
    "level": 1,
    "metadata": Object {
      "annotations": Object {},
      "enum": undefined,
      "id": "random-id",
      "name": "_name$",
      "path": Array [
        "patternProperties",
        "_name$",
      ],
      "pattern": true,
      "type": "string",
      "validations": Object {},
    },
    "name": "",
  },
]
`;

exports[`renderSchema util should match ref/original.json 1`] = `
Array [
  Object {
    "canHaveChildren": true,
    "id": "random-id",
    "level": 0,
    "metadata": Object {
      "additionalProperties": undefined,
      "annotations": Object {},
      "enum": undefined,
      "id": "random-id",
      "path": Array [],
      "patternProperties": undefined,
      "properties": Object {
        "billing_address": Object {
          "$ref": "#/definitions/address",
        },
        "shipping_address": Object {
          "$ref": "#/definitions/address",
        },
      },
      "type": "object",
      "validations": Object {},
    },
    "name": "",
  },
  Object {
    "canHaveChildren": true,
    "id": "random-id",
    "level": 1,
    "metadata": Object {
      "additionalProperties": undefined,
      "annotations": Object {},
      "enum": undefined,
      "id": "random-id",
      "inheritedFrom": "#/definitions/address",
      "name": "billing_address",
      "path": Array [
        "properties",
        "billing_address",
      ],
      "patternProperties": undefined,
      "properties": Object {
        "city": Object {
          "type": "string",
        },
        "state": Object {
          "type": "string",
        },
        "street_address": Object {
          "type": "string",
        },
      },
      "required": false,
      "type": "object",
      "validations": Object {},
    },
    "name": "",
  },
  Object {
    "id": "random-id",
    "level": 2,
    "metadata": Object {
      "annotations": Object {},
      "enum": undefined,
      "id": "random-id",
      "name": "street_address",
      "path": Array [
        "properties",
        "billing_address",
        "properties",
        "street_address",
      ],
      "required": true,
      "type": "string",
      "validations": Object {},
    },
    "name": "",
  },
  Object {
    "id": "random-id",
    "level": 2,
    "metadata": Object {
      "annotations": Object {},
      "enum": undefined,
      "id": "random-id",
      "name": "city",
      "path": Array [
        "properties",
        "billing_address",
        "properties",
        "city",
      ],
      "required": true,
      "type": "string",
      "validations": Object {},
    },
    "name": "",
  },
  Object {
    "id": "random-id",
    "level": 2,
    "metadata": Object {
      "annotations": Object {},
      "enum": undefined,
      "id": "random-id",
      "name": "state",
      "path": Array [
        "properties",
        "billing_address",
        "properties",
        "state",
      ],
      "required": true,
      "type": "string",
      "validations": Object {},
    },
    "name": "",
  },
  Object {
    "canHaveChildren": true,
    "id": "random-id",
    "level": 1,
    "metadata": Object {
      "additionalProperties": undefined,
      "annotations": Object {},
      "enum": undefined,
      "id": "random-id",
      "inheritedFrom": "#/definitions/address",
      "name": "shipping_address",
      "path": Array [
        "properties",
        "shipping_address",
      ],
      "patternProperties": undefined,
      "properties": Object {
        "city": Object {
          "type": "string",
        },
        "state": Object {
          "type": "string",
        },
        "street_address": Object {
          "type": "string",
        },
      },
      "required": false,
      "type": "object",
      "validations": Object {},
    },
    "name": "",
  },
  Object {
    "id": "random-id",
    "level": 2,
    "metadata": Object {
      "annotations": Object {},
      "enum": undefined,
      "id": "random-id",
      "name": "street_address",
      "path": Array [
        "properties",
        "shipping_address",
        "properties",
        "street_address",
      ],
      "required": true,
      "type": "string",
      "validations": Object {},
    },
    "name": "",
  },
  Object {
    "id": "random-id",
    "level": 2,
    "metadata": Object {
      "annotations": Object {},
      "enum": undefined,
      "id": "random-id",
      "name": "city",
      "path": Array [
        "properties",
        "shipping_address",
        "properties",
        "city",
      ],
      "required": true,
      "type": "string",
      "validations": Object {},
    },
    "name": "",
  },
  Object {
    "id": "random-id",
    "level": 2,
    "metadata": Object {
      "annotations": Object {},
      "enum": undefined,
      "id": "random-id",
      "name": "state",
      "path": Array [
        "properties",
        "shipping_address",
        "properties",
        "state",
      ],
      "required": true,
      "type": "string",
      "validations": Object {},
    },
    "name": "",
  },
]
`;
